File: skills/decision-free-specs/languages/typescript/inventory.mjs
#!/usr/bin/env node
// Symbol-level inventory of TS/TSX files, for writing decision-free refactor specs.
// Usage (from the TARGET repo's root):
//   node <path-to-skill>/languages/typescript/inventory.mjs <file.tsx> [...]
// Output: markdown to stdout.
//
// Replaces the expensive "agent reads the file and lists symbols" planning step:
// everything a spec's symbol map needs (names, kinds, line ranges, export status,
// component detection, hooks, API calls, dynamic classNames, importers) comes from
// the AST. Uses the TARGET repo's own `typescript` package — no dependencies here.

import { execSync } from "node:child_process";
import { readdirSync, readFileSync, statSync } from "node:fs";
import { createRequire } from "node:module";
import path from "node:path";

let ts;
try {
	ts = createRequire(path.join(process.cwd(), "package.json"))("typescript");
} catch {
	console.error(
		"Could not resolve 'typescript' from the current directory.\n" +
			"Run this script from the root of the repo you are inventorying\n" +
			"(it must have typescript installed, e.g. `npm i -D typescript`).",
	);
	process.exit(1);
}

const files = process.argv.slice(2);
if (files.length === 0) {
	console.error("usage: node inventory.mjs <file.tsx> [...]   (run from the target repo root)");
	process.exit(1);
}

function listSourceFiles() {
	try {
		return execSync('git ls-files "*.ts" "*.tsx"', { encoding: "utf8" }).split("\n").filter(Boolean);
	} catch {
		const out = [];
		const walk = (dir) => {
			for (const entry of readdirSync(dir)) {
				if (entry === "node_modules" || entry.startsWith(".")) continue;
				const p = path.join(dir, entry);
				if (statSync(p).isDirectory()) walk(p);
				else if (/\.(ts|tsx)$/.test(entry)) out.push(p);
			}
		};
		walk(".");
		return out;
	}
}

function lineOf(sf, pos) {
	return sf.getLineAndCharacterOfPosition(pos).line + 1;
}

function isExported(node) {
	return (ts.getModifiers?.(node) ?? node.modifiers ?? []).some((m) => m.kind === ts.SyntaxKind.ExportKeyword);
}

function containsJsx(node) {
	let found = false;
	const visit = (n) => {
		if (found) return;
		if (ts.isJsxElement(n) || ts.isJsxSelfClosingElement(n) || ts.isJsxFragment(n)) {
			found = true;
			return;
		}
		ts.forEachChild(n, visit);
	};
	visit(node);
	return found;
}

function hookCalls(node) {
	const hooks = [];
	const visit = (n) => {
		if (ts.isVariableDeclaration(n) && n.initializer && ts.isCallExpression(n.initializer)) {
			const callee = n.initializer.expression.getText();
			if (/^(useState|useRef|useMemo|useCallback|useSyncExternalStore)$/.test(callee)) {
				hooks.push(`${callee}: ${n.name.getText()}`);
			}
		}
		ts.forEachChild(n, visit);
	};
	visit(node);
	return hooks;
}

function describeDeclaration(sf, stmt) {
	const rows = [];
	const span = (node) => `${lineOf(sf, node.getStart(sf))}-${lineOf(sf, node.getEnd())}`;
	if (ts.isTypeAliasDeclaration(stmt) || ts.isInterfaceDeclaration(stmt)) {
		rows.push({ kind: "type", name: stmt.name.text, lines: span(stmt), exported: isExported(stmt), node: stmt });
	} else if (ts.isFunctionDeclaration(stmt) && stmt.name) {
		const component = /^[A-Z]/.test(stmt.name.text) && containsJsx(stmt);
		rows.push({
			kind: component ? "component" : "function",
			name: stmt.name.text,
			lines: span(stmt),
			exported: isExported(stmt),
			node: stmt,
		});
	} else if (ts.isVariableStatement(stmt)) {
		for (const decl of stmt.declarationList.declarations) {
			const name = decl.name.getText();
			const init = decl.initializer;
			const arrow = init && (ts.isArrowFunction(init) || ts.isFunctionExpression(init));
			const component = arrow && /^[A-Z]/.test(name) && containsJsx(init);
			rows.push({
				kind: component ? "component" : arrow ? "function" : "const",
				name,
				lines: span(stmt),
				exported: isExported(stmt),
				node: stmt,
			});
		}
	} else if (ts.isClassDeclaration(stmt) && stmt.name) {
		rows.push({ kind: "class", name: stmt.name.text, lines: span(stmt), exported: isExported(stmt), node: stmt });
	}
	return rows;
}

function grepWithLines(sf, text, regex) {
	const out = [];
	for (const match of text.matchAll(regex)) {
		out.push({ line: lineOf(sf, match.index), text: match[0].slice(0, 120) });
	}
	return out;
}

function importersOf(targetFile, allFiles) {
	const base = path.basename(targetFile).replace(/\.(ts|tsx)$/, "");
	const hits = [];
	for (const f of allFiles) {
		if (path.resolve(f) === path.resolve(targetFile)) continue;
		const text = readFileSync(f, "utf8");
		const re = new RegExp(`import\\s+(type\\s+)?\\{([^}]*)\\}\\s+from\\s+"[^"]*/${base}"`, "g");
		for (const m of text.matchAll(re)) {
			hits.push({ file: f, names: m[2].replace(/\s+/g, " ").trim() });
		}
		if (new RegExp(`from\\s+"[^"]*/${base}"`).test(text) && !hits.some((h) => h.file === f)) {
			hits.push({ file: f, names: "(default or side-effect import)" });
		}
	}
	return hits;
}

const allSrcFiles = listSourceFiles();

for (const file of files) {
	const text = readFileSync(file, "utf8");
	const sf = ts.createSourceFile(file, text, ts.ScriptTarget.Latest, true, ts.ScriptKind.TSX);
	const lineCount = text.split("\n").length;

	console.log(`# Inventory: ${file}\n`);
	console.log(`Total lines: ${lineCount}\n`);

	console.log("## Imports\n");
	for (const stmt of sf.statements) {
		if (ts.isImportDeclaration(stmt)) console.log(`- \`${stmt.getText(sf).replace(/\s+/g, " ")}\``);
	}

	console.log("\n## Top-level declarations (in order)\n");
	console.log("| Kind | Name | Lines | Exported |");
	console.log("|---|---|---|---|");
	const declarations = [];
	for (const stmt of sf.statements) {
		for (const row of describeDeclaration(sf, stmt)) {
			declarations.push(row);
			console.log(`| ${row.kind} | \`${row.name}\` | ${row.lines} | ${row.exported ? "yes" : "no"} |`);
		}
	}

	for (const row of declarations) {
		if (row.kind !== "component") continue;
		const hooks = hookCalls(row.node);
		if (hooks.length) {
			console.log(`\n### Hooks in \`${row.name}\`\n`);
			for (const h of hooks) console.log(`- ${h}`);
		}
	}

	const apiCalls = grepWithLines(sf, text, /\bclient\.(GET|POST|PATCH|PUT|DELETE)\(\s*"([^"]+)"/g);
	console.log("\n## API calls\n");
	if (apiCalls.length === 0) console.log("(none)");
	for (const c of apiCalls) console.log(`- L${c.line}: \`${c.text}\``);

	const sse = grepWithLines(sf, text, /new EventSource\([^)]*\)/g);
	if (sse.length) {
		console.log("\n## EventSource\n");
		for (const c of sse) console.log(`- L${c.line}: \`${c.text}\``);
	}

	const dynClass = grepWithLines(sf, text, /className=\{`[^`]*\$\{[^`]*`\}/g);
	console.log("\n## Dynamic classNames (grep-invisible — never delete their CSS)\n");
	if (dynClass.length === 0) console.log("(none)");
	for (const c of dynClass) console.log(`- L${c.line}: \`${c.text}\``);

	const importers = importersOf(file, allSrcFiles);
	console.log("\n## Imported by (load-bearing exports — keep compatible)\n");
	if (importers.length === 0) console.log("(no other source file imports this module)");
	for (const i of importers) console.log(`- \`${i.file}\` imports: ${i.names}`);

	console.log("\n---\n");
}

File: skills/decision-free-specs/languages/python/inventory.py
#!/usr/bin/env python3
"""Symbol-level inventory of Python files, for writing decision-free refactor specs.

Usage (from the TARGET repo's root):
    python3 <path-to-skill>/languages/python/inventory.py <file.py> [...]

Output: markdown to stdout, per the contract in languages/README.md.
Stdlib only (ast); no dependencies.
"""

import ast
import re
import subprocess
import sys
from pathlib import Path


def list_source_files():
    try:
        out = subprocess.run(
            ["git", "ls-files", "*.py"], capture_output=True, text=True, check=True
        ).stdout
        return [line for line in out.splitlines() if line]
    except Exception:
        return [str(p) for p in Path(".").rglob("*.py") if "node_modules" not in p.parts and not any(part.startswith(".") for part in p.parts)]


def exported_name(name, dunder_all):
    if dunder_all is not None:
        return name in dunder_all
    return not name.startswith("_")


def read_dunder_all(tree):
    for node in tree.body:
        if isinstance(node, ast.Assign):
            for target in node.targets:
                if isinstance(target, ast.Name) and target.id == "__all__":
                    try:
                        return [elt.value for elt in node.value.elts if isinstance(elt, ast.Constant)]
                    except AttributeError:
                        return None
    return None


def decorators_of(node):
    return [ast.unparse(d) for d in getattr(node, "decorator_list", [])]


def describe(tree, dunder_all):
    rows = []
    for node in tree.body:
        span = f"{node.lineno}-{node.end_lineno}"
        if isinstance(node, ast.ClassDef):
            rows.append(("class", node.name, span, exported_name(node.name, dunder_all), node))
        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            rows.append(("function", node.name, span, exported_name(node.name, dunder_all), node))
        elif isinstance(node, ast.Assign):
            for target in node.targets:
                if isinstance(target, ast.Name):
                    rows.append(("const", target.id, span, exported_name(target.id, dunder_all), node))
        elif isinstance(node, ast.AnnAssign) and isinstance(node.target, ast.Name):
            rows.append(("const", node.target.id, span, exported_name(node.target.id, dunder_all), node))
        elif hasattr(ast, "TypeAlias") and isinstance(node, ast.TypeAlias):
            rows.append(("type", ast.unparse(node.name), span, True, node))
    return rows


def dynamic_access(tree):
    """Flag dynamic symbol access — the grep-invisible trap in Python."""
    hits = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Call) and isinstance(node.func, ast.Name) and node.func.id in ("getattr", "globals", "vars", "__import__"):
            hits.append((node.lineno, ast.unparse(node)[:120]))
        elif isinstance(node, ast.Call) and isinstance(node.func, ast.Attribute) and node.func.attr == "import_module":
            hits.append((node.lineno, ast.unparse(node)[:120]))
    return hits


def importers_of(target_file, all_files):
    stem = Path(target_file).stem
    if stem == "__init__":
        stem = Path(target_file).parent.name
    hits = []
    from_re = re.compile(rf"^\s*from\s+[\w.]*\b{re.escape(stem)}\b\s+import\s+(.+)$", re.M)
    import_re = re.compile(rf"^\s*import\s+[\w.]*\b{re.escape(stem)}\b", re.M)
    for f in all_files:
        if Path(f).resolve() == Path(target_file).resolve():
            continue
        try:
            text = Path(f).read_text(encoding="utf-8")
        except Exception:
            continue
        for m in from_re.finditer(text):
            hits.append((f, m.group(1).strip()[:100]))
        if import_re.search(text) and not any(h[0] == f for h in hits):
            hits.append((f, "(module import)"))
    return hits


def main():
    files = sys.argv[1:]
    if not files:
        print("usage: inventory.py <file.py> [...]   (run from the target repo root)", file=sys.stderr)
        sys.exit(1)
    all_files = list_source_files()

    for file in files:
        text = Path(file).read_text(encoding="utf-8")
        tree = ast.parse(text, filename=file)
        dunder_all = read_dunder_all(tree)

        print(f"# Inventory: {file}\n")
        print(f"Total lines: {len(text.splitlines())}\n")

        print("## Imports\n")
        for node in tree.body:
            if isinstance(node, (ast.Import, ast.ImportFrom)):
                print(f"- `{ast.unparse(node)}`")

        print("\n## Top-level declarations (in order)\n")
        print("| Kind | Name | Lines | Exported |")
        print("|---|---|---|---|")
        rows = describe(tree, dunder_all)
        for kind, name, span, exported, _node in rows:
            print(f"| {kind} | `{name}` | {span} | {'yes' if exported else 'no'} |")

        if dunder_all is not None:
            print(f"\n## `__all__`\n\n`{dunder_all}` — exported means listed here.")

        for kind, name, _span, _exported, node in rows:
            if kind == "class":
                methods = [n.name for n in node.body if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef))]
                if methods:
                    print(f"\n### Methods in `{name}`\n")
                    for m in methods:
                        print(f"- {m}")
            decs = decorators_of(node) if kind in ("class", "function") else []
            if decs:
                print(f"\n### Decorators on `{name}`\n")
                for d in decs:
                    print(f"- `@{d}` (may register the symbol by module path — see NOTES.md)")

        dyn = dynamic_access(tree)
        print("\n## Dynamic symbol access (grep-invisible — verify before moving)\n")
        if not dyn:
            print("(none)")
        for line, snippet in dyn:
            print(f"- L{line}: `{snippet}`")

        importers = importers_of(file, all_files)
        print("\n## Imported by (load-bearing exports — keep compatible)\n")
        if not importers:
            print("(no other source file imports this module)")
        for f, names in importers:
            print(f"- `{f}` imports: {names}")

        print("\n---\n")


if __name__ == "__main__":
    main()
